src/ahp/constraint_systems.rs

```diff
@@ -54,190 +33,63 @@ fn balance_matrices<F: Field>(a_matrix: &mut Matrix<F>, b_matrix: &mut Matrix<F>
     }
 }
 
-/// Stores constraints during index generation.
-pub(crate) struct IndexerConstraintSystem<F: Field> {
-    pub(crate) num_input_variables: usize,
-    pub(crate) num_witness_variables: usize,
-    pub(crate) num_constraints: usize,
-    pub(crate) a: Vec<Vec<(F, VarIndex)>>,
-    pub(crate) b: Vec<Vec<(F, VarIndex)>>,
-    pub(crate) c: Vec<Vec<(F, VarIndex)>>,
-
-    pub(crate) a_matrix: Option<Matrix<F>>,
-    pub(crate) b_matrix: Option<Matrix<F>>,
-    pub(crate) c_matrix: Option<Matrix<F>>,
+pub(crate) fn num_non_zero<F: PrimeField>(cs: ConstraintSystemRef<F>) -> usize {
+    let matrices = cs.to_matrices().unwrap();
```

Comment:
This is kind of an expensive operation, so I think it would make sense to pass the matrices by reference?